src/jit/function-builder.cc

```diff
@@ -246,116 +246,145 @@ bool FunctionBuilder::buildIL() {
   setVMState(new TR::VirtualMachineState());
 
   const uint8_t* istream = thread_->GetIstream();
+  const uint8_t* pc = &istream[fn_->offset];
+  auto* state = new WabtState();
 
-  workItems_.emplace_back(OrphanBytecodeBuilder(0, const_cast<char*>(ReadOpcodeAt(&istream[fn_->offset]).GetName())),
-                          &istream[fn_->offset]);
+  SetUpLocals(this, &pc, &state->stack);
+  setVMState(state);
+
+  workItems_.emplace_back(OrphanBytecodeBuilder(0, const_cast<char*>(ReadOpcodeAt(pc).GetName())),
+                          pc);
   AppendBuilder(workItems_[0].builder);
 
   int32_t next_index;
 
   while ((next_index = GetNextBytecodeFromWorklist()) != -1) {
     auto& work_item = workItems_[next_index];
 
-    if (!Emit(work_item.builder, istream, work_item.pc))
+    if (!Emit(work_item.builder, static_cast<WabtState*>(work_item.builder->vmState())->stack, istream, work_item.pc))
```

Comment:
Would it not be better to just pass `state->stack`?

Reply:
At this point in the function `state` is stale (it refers to the state right before the first `BytecodeBuilder`), so directly replacing this with `state->stack` would be incorrect. I agree that this isn't very pretty, though. I think I'll go ahead and add a reload of `state` from `work_item.builder->vmState()` and then use `state->stack` to reduce the amount of stuff going on in this line, as this could quickly get out-of-hand if we start adding more state.